commons/src/main/java/io/aiven/kafka/tieredstorage/commons/SegmentManifestProvider.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.commons;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+
+import io.aiven.kafka.tieredstorage.commons.manifest.SegmentManifest;
+import io.aiven.kafka.tieredstorage.commons.storage.FileFetcher;
+import io.aiven.kafka.tieredstorage.commons.storage.StorageBackEndException;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+class SegmentManifestProvider {
+    private final ObjectKey objectKey;
+    private final Cache<String, SegmentManifest> cache;
+    private final FileFetcher fileFetcher;
+    private final ObjectMapper mapper;
+
+    /**
+     * @param maxCacheSize the max cache size (in items) or {@code null} if the cache is unbounded.
+     * @param cacheRetention the retention time of items in the cache or {@code null} if infinite retention.
+     */
+    SegmentManifestProvider(final ObjectKey objectKey,
+                            final Long maxCacheSize,
+                            final Duration cacheRetention,
+                            final FileFetcher fileFetcher,
+                            final ObjectMapper mapper) {
+        this.objectKey = objectKey;
+        final var cacheBuilder = Caffeine.newBuilder();
+        if (maxCacheSize != null) {
+            cacheBuilder.maximumSize(maxCacheSize);
+        }
+        if (cacheRetention != null) {
```

Comment:
nit: should we use optional instead?

Reply:
I guess optional does not really make sense here, rather same as above - check for -1 here instead of config maybe.

Reply:
Does not seem to be a good idea to have an Optional constructor argument. as(again) mentioned in another thread I will try to address this in a different way a little later.

Reply:
Well, `Optional` is not recommended as a parameter in general case because of some valid reasons. However, values coming from configuration seem like a pretty valid special case where `Optional`s are just fine.